### PR TITLE
Add npm package release workflow

### LINKED ISSUE #10251: Add workflow for releasing npm packages
## Task Description

Add GitHub workflow for publishing packages regularly

### PR DESCRIPTION
**Pull request by swissspidy**
Supersedes exploratory PR #9863

This PR implements changes necessary for publishing our packages intended for public use to the npm package registry.

This is highly relevant for the decoupling effort, because we want to allow others to easily consume this code for use on their platforms.

**In this PR**

* Move migration CLI script out of `src`
* Update `package.json` configurations for those public packages (version, description, `npmignore`, etc.
* Explore automated package publishing mechanism  

**Note:** Using incremental versions like `v0.1.<YYYYMMDDHHMM` for the time being, , eg. `v0.1.202111302140`. We can look into better versioning in the future.

**To-do**

* [ ] Use [dist-tags](https://docs.npmjs.com/cli/v8/commands/npm-dist-tag)? `latest` is added automatically, but maybe we want `dev`, `canary` or something like that in the future.
* [ ] Only push the committed version change to `main` when actually on `main` branch and when dry-run succeeds
* [ ] Push version change as the Google for Creators Bot account

**Note:** Right now, code review is best done by looking at the individual commits. I might split this up into multiple PRs later.

This change allows publishing packages under the `@googleforcreators` scope, which at the moment would only include these public packages:

* `@googleforcreators/animation`
* `@googleforcreators/dashboard`
* `@googleforcreators/date`
* `@googleforcreators/design-system`
* `@googleforcreators/fonts`
* `@googleforcreators/i18n`
* `@googleforcreators/media`
* `@googleforcreators/migration`
* `@googleforcreators/patterns`
* `@googleforcreators/react`
* `@googleforcreators/stickers`
* `@googleforcreators/story-editor`
* `@googleforcreators/templates` (see also #7196)
* `@googleforcreators/text-sets`
* `@googleforcreators/tracking`
* `@googleforcreators/units`

**How this has been tested**

```bash
# In this project
$ bash bin/setup-local-npm-registry.sh
$ npm version --workspaces --no-git-tag-version "0.1.$(date -u +%Y%m%d%H%M)"
$ npm --registry http://localhost:4873/ login
$ npm --registry http://localhost:4873/ --workspaces publish

# In another project
$ npm --registry http://localhost:4873 install --save @googleforcreators/story-editor @googleforcreators/dashboard # ... and more
$ npm list
test@1.0.0 /path/to/project
├── @googleforcreators/dashboard@0.1.202112082130
├── @googleforcreators/date@0.1.202112082130
├── @googleforcreators/design-system@0.1.202112082130
├── @googleforcreators/i18n@0.1.202112082130
├── @googleforcreators/media@0.1.202112082130
├── @googleforcreators/patterns@0.1.202112082130
├── @googleforcreators/react@0.1.202112082130
├── @googleforcreators/story-editor@0.1.202112082130
└── @googleforcreators/tracking@0.1.202112082130
```

On my local test registry the result looks like this:

Screenshot 2021-12-01 at 16 43 30

Fixes #10251

### CODE REVIEW

**Comment by sayedtaqui** [comment]
Trying to use the local registry packages, to spin up a minimal story editor, but I get webpack errors for some loader when running it in another project. Any clue what these could be 🤔 

image

**Comment by mohdsayed** [comment]
Trying to spin up a story editor with local registry packages, using create-react-app. There seems to be some issue to the exports. 

image

**Comment by swissspidy** [comment]
Ah, good catch.

Does it work if you—after installing the package locally— manually modify `node_modules/@googleforcreators/story-editor/package.json` and remove `"type": "module"`?

**Comment by mohdsayed** [comment]
That solved the export issue. Now another one for babel.

( Tried creating[ Gutenberg playground](https://github.com/WordPress/gutenberg/blob/trunk/storybook/stories/playground/index.js) with the same setup and its working without any issue  )

image

**Comment by swissspidy** [comment]
Clearly I haven't spent time testing testing these packages beyond just `npm install`ing them 🤦 

In hindsight it's quite obvious. _Of course_ we need to build packages, including JSX and styled components etc. before publishing them.

I'll look into it

File: bin/setup-local-npm-registry.sh [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
```
Comment (mohdsayed) [inline bin/setup-local-npm-registry.sh:1]: Should we probably add some npm scripts to execute these files in `package.json`  like we have commands for other `./bin` files?

**Review by mohdsayed — APPROVED** [review]

File: bin/setup-local-npm-registry.sh (hunk 1 above)
Comment (swissspidy) [reply bin/setup-local-npm-registry.sh:1]: Yeah why not 👍  Will add

**Pull request merged**